firebase_admin/db.py

```diff
@@ -191,26 +181,72 @@ def update(self, value):
             raise ValueError('Dictionary must not contain None keys or values.')
         self._client.request_oneway('patch', self._add_suffix(), json=value, params='print=silent')
 
+    def update_with_etag(self, value, etag):
+        """Updates the specified child keys of this Reference to the provided values
+            and uses ETag to make sure data is up to date.
+        Args:
+            value: A dictionary containing the child keys to update, and their new values.
+            etag: ETag value for the Reference.
+        Returns:
+            value: None if the update is successful, otherwise the current ETag of the reference
+            and a snapshot of the data in the database.
+        Raises:
+            ValueError: If value is empty or not a dictionary, or if etag is not a string.
+        """
+        if not value or not isinstance(value, dict):
+            raise ValueError('Value argument must be a non-empty dictionary.')
+        if None in value.keys() or None in value.values():
+            raise ValueError('Dictionary must not contain None keys or values.')
+        if not isinstance(etag, str):
+            raise ValueError('ETag must be a string.')
+
+        try:
+            self._client.request_oneway('put', self._add_suffix(), json=value,
+                                        headers={'if-match': etag})
+        except ApiCallError as error:
+            detail = error.detail
+            snapshot = detail.response.json()
+            etag = detail.response.headers['ETag']
+            return etag, snapshot
+
     def delete(self):
         """Deleted this node from the database.
-
         Raises:
           ApiCallError: If an error occurs while communicating with the remote database server.
         """
         self._client.request_oneway('delete', self._add_suffix())
 
+    def transaction(self, transaction_update):
+        """Write to database using a transaction.
+        Args:
+            transaction_update: function that takes in current database data as a parameter.
+        Raises:
+            ValueError: If transaction_update is not a function.
+
+        """
+        if not callable(transaction_update):
+            raise ValueError('transaction_update must be a function.')
+
+        tries = 0
+        etag, data = self.get_with_etag()
+        val = transaction_update(data)
+        while tries < _TRANSACTION_MAX_RETRIES:
+            resp = self.update_with_etag(val, etag)
+            if resp is None:
+                break
+            else:
+                etag, data = resp
+                val = transaction_update(data)
+                tries += 1
```

Comment:
We can restructure this while loop a little:

```
etag, data = self.get_with_etag()
while tries < _TRANSACTION_MAX_RETRIES:
    val, should_retry = transaction_update(data)
    if not should_retry:
        break
    resp = self.update_with_etag(val, etag)
    if resp is None:
        break
    else:
        etag, data = resp
        tries += 1
```
This also lets the user specify when to stop retrying in transaction_update. Note that this code will raise an exception if transaction_update does not return a tuple. We probably want to catch that exception and raise a one with a context specific message in its place.

Note that should_retry need not be a boolean. We could either assert that it is a boolean and raise a clear exception to prevent users from accidentally passing unintended values through as should_retry, or we could leave it as is, assuming that in some cases a user may be passing through things like an empty list versus a non-empty list as their retry criteria.